File: full-vineflower/me/ag2s/umdlib/tool/WrapOutputStream.java
package me.ag2s.umdlib.tool;

import java.io.IOException;
import java.io.OutputStream;

public class WrapOutputStream extends OutputStream {
   private OutputStream os;
   private int written;

   public WrapOutputStream(OutputStream os) {
      this.os = os;
   }

   private void incCount(int value) {
      int temp = this.written + value;
      if (temp < 0) {
         temp = Integer.MAX_VALUE;
      }

      this.written = temp;
   }

   public void writeInt(int v) throws IOException {
      this.os.write(v >>> 0 & 0xFF);
      this.os.write(v >>> 8 & 0xFF);
      this.os.write(v >>> 16 & 0xFF);
      this.os.write(v >>> 24 & 0xFF);
      this.incCount(4);
   }

   public void writeByte(byte b) throws IOException {
      this.write(b);
   }

   public void writeByte(int n) throws IOException {
      this.write(n);
   }

   public void writeBytes(byte... bytes) throws IOException {
      this.write(bytes);
   }

   public void writeBytes(int... vals) throws IOException {
      for (int v : vals) {
         this.write(v);
      }
   }

   @Override
   public void write(byte[] b, int off, int len) throws IOException {
      this.os.write(b, off, len);
      this.incCount(len);
   }

   @Override
   public void write(byte[] b) throws IOException {
      this.os.write(b);
      this.incCount(b.length);
   }

   @Override
   public void write(int b) throws IOException {
      this.os.write(b);
      this.incCount(1);
   }

   @Override
   public void close() throws IOException {
      this.os.close();
   }

   @Override
   public void flush() throws IOException {
      this.os.flush();
   }

   @Override
   public boolean equals(Object obj) {
      return this.os.equals(obj);
   }

   @Override
   public int hashCode() {
      return this.os.hashCode();
   }

   @Override
   public String toString() {
      return this.os.toString();
   }

   public int getWritten() {
      return this.written;
   }
}

File: best-of-3/business/me/ag2s/umdlib/tool/WrapOutputStream.kt
package me.ag2s.umdlib.tool

import java.io.IOException
import java.io.OutputStream

/** Little-endian counting output stream used by UMD builders. */
class WrapOutputStream(private val os: OutputStream) : OutputStream() {
    var written: Int = 0
        private set

    private fun inc(n: Int) {
        val t = written + n
        written = if (t < 0) Int.MAX_VALUE else t
    }

    fun writeInt(v: Int) {
        os.write(v ushr 0 and 0xFF)
        os.write(v ushr 8 and 0xFF)
        os.write(v ushr 16 and 0xFF)
        os.write(v ushr 24 and 0xFF)
        inc(4)
    }

    fun writeByte(b: Byte) = write(b.toInt() and 0xFF)

    fun writeByte(n: Int) = write(n and 0xFF)

    fun writeBytes(vararg bytes: Byte) {
        write(bytes)
    }

    fun writeBytes(vararg vals: Int) {
        for (v in vals) write(v and 0xFF)
    }

    @Throws(IOException::class)
    override fun write(b: ByteArray, off: Int, len: Int) {
        os.write(b, off, len)
        inc(len)
    }

    @Throws(IOException::class)
    override fun write(b: ByteArray) {
        os.write(b)
        inc(b.size)
    }

    @Throws(IOException::class)
    override fun write(b: Int) {
        os.write(b)
        inc(1)
    }

    @Throws(IOException::class)
    override fun close() = os.close()

    @Throws(IOException::class)
    override fun flush() = os.flush()
}
